File: AVCameExpert WatchKit App Extension/InterfaceController.m
//
//  InterfaceController.m
//  AVCameExpert WatchKit App Extension
//
//  Created by Sébastien Luneau on 18/02/2016.
//  Copyright © 2016 Matchpix. All rights reserved.
//

@import WatchConnectivity;
#import "InterfaceController.h"


@interface InterfaceController()<WCSessionDelegate>
@property (unsafe_unretained, nonatomic) IBOutlet WKInterfaceImage *imageView;
@property (unsafe_unretained, nonatomic) IBOutlet WKInterfaceLabel *startLabel;
@property (unsafe_unretained, nonatomic) IBOutlet WKInterfaceGroup *mainGroup;

@end


@implementation InterfaceController

- (void)awakeWithContext:(id)context {
    [super awakeWithContext:context];

    // Configure interface objects here.
    if ([WCSession isSupported]) {
        WCSession *session = [WCSession defaultSession];
        session.delegate = self;
        [session activateSession];
    }
}

- (void)willActivate {
    // This method is called when watch view controller is about to be visible to user
    [super willActivate];
    [_startLabel setHidden:NO];
    [_mainGroup setHidden:YES];
}

- (void)didDeactivate {
    // This method is called when watch view controller is no longer visible
    
    [_startLabel setHidden:NO];
    [_mainGroup setHidden:YES];
    [super didDeactivate];
}
- (IBAction)grabAction {
    if ([[WCSession defaultSession] isReachable]) {
    
    NSDictionary *applicationDict = @{@"action":@"grab"};
    [[WCSession defaultSession] sendMessage:applicationDict
                               replyHandler:^(NSDictionary *replyHandler) {
                                   
                               }
                               errorHandler:^(NSError *error) {
                                   
                               }
     ];
    }
}
- (void)session:(WCSession *)session didReceiveMessageData:(NSData *)messageData
   replyHandler:(void (^)(NSData *replyMessageData))replyHandler{
    
    [_startLabel setHidden:YES];
    [_mainGroup setHidden:NO];

    UIImage* image = [UIImage imageWithData:messageData];
    dispatch_async( dispatch_get_main_queue(), ^{
    [_imageView setImage:image];
    });
    //replyHandler([[NSData alloc] init]);
    
}
-(void)session:(WCSession *)session didReceiveMessageData:(NSData *)messageData{
    
    [_startLabel setHidden:YES];
    [_mainGroup setHidden:NO];

    UIImage* image = [UIImage imageWithData:messageData];
    dispatch_async( dispatch_get_main_queue(), ^{
        [_imageView setImage:image];
    });
}
- (IBAction)switchAction {
    if ([[WCSession defaultSession] isReachable]) {
        NSDictionary *applicationDict = @{@"action":@"switch"};
        [[WCSession defaultSession] sendMessage:applicationDict
                                   replyHandler:^(NSDictionary *replyHandler) {
                                       
                                   }
                                   errorHandler:^(NSError *error) {
                                       
                                   }
         ];

    }
}
@end

File: AVCamExpert WatchKit Extension/InterfaceController.swift
//
//  InterfaceController.swift
//  test WatchKit Extension
//
//  Created by Sébastien Luneau on 18/02/2016.
//  Copyright © 2016 MatchPix SARL. All rights reserved.
//

import WatchKit
import WatchConnectivity
import Foundation

class InterfaceController: WKInterfaceController, WCSessionDelegate {
    /** Called when the session has completed activation. If session state is WCSessionActivationStateNotActivated there will be an error with more details. */
    @available(watchOS 2.2, *)
    public func session(_ session: WCSession, activationDidCompleteWith activationState: WCSessionActivationState, error: Error?) {
        
    }

    @IBOutlet var imageView: WKInterfaceImage!
    
    @IBOutlet var mainGroup: WKInterfaceGroup!
    @IBOutlet var startLabel: WKInterfaceLabel!
    override func awake(withContext context: Any?) {
        super.awake(withContext: context)
        
        // Configure interface objects here.
        if (WCSession.isSupported()){
            let session = WCSession.default()
            session.delegate = self
            session.activate()
        }
    }
    
    override func willActivate() {
        // This method is called when watch view controller is about to be visible to user
        super.willActivate()
        mainGroup.setHidden(true)
        startLabel.setHidden(false)
    }
    
    override func didDeactivate() {
        // This method is called when watch view controller is no longer visible
        super.didDeactivate()
        mainGroup.setHidden(true)
        startLabel.setHidden(false)
        
    }
    
    // MARK: -  WatchConnectivity delegates
    func session(_ session: WCSession, didReceiveMessageData messageData: Data) {
        mainGroup.setHidden(false)
        startLabel.setHidden(true)
        
        DispatchQueue.main.async { () -> Void in
            self.imageView.setImageData(messageData)
        }
        
    }
    func session(_ session: WCSession, didReceiveMessageData messageData: Data, replyHandler: @escaping (Data) -> Void) {
        mainGroup.setHidden(false)
        startLabel.setHidden(true)
        
        DispatchQueue.main.async { () -> Void in
            self.imageView.setImageData(messageData)
        }
    }
    // MARK: -  actions
    @IBAction func grabAction() {
        if (WCSession.default().isReachable){
            let messageDict = ["action":"grab"]
            WCSession.default().sendMessage(messageDict, replyHandler: { (_: [String : Any]) in
                // do something            
            }, errorHandler: { (Error) in
                // do something
            })

            
        }
    }
    @IBAction func switchAction() {
        if (WCSession.default().isReachable){
            let messageDict = ["action":"switch"]
            
            WCSession.default().sendMessage(messageDict, replyHandler: { (_: [String : Any]) in
                // do something
            }, errorHandler: { (Error) in
                // do something
            })
            
        }
    }
    
}
